src/Perl6/bootstrap.c/BOOTSTRAP.nqp

```diff
@@ -4055,93 +4054,114 @@ nqp::sethllconfig('Raku', nqp::hash(
     },
     'exit_handler', -> $coderef, $resultish {
         unless nqp::p6inpre() {
-            my %phasers :=
-              nqp::getattr(nqp::getcodeobj($coderef),Block,'$!phasers');
-            my @leaves := nqp::atkey(%phasers, '!LEAVE-ORDER');
-            my @posts  := nqp::atkey(%phasers, 'POST');
-            my @exceptions;
-            unless nqp::isnull(@leaves) {
-
-                # only have a single LEAVEish phaser, so no frills needed
-                if nqp::elems(@leaves) == 1 && nqp::elems(%phasers) == 1 {
+            # when we get here, we assume the $!phasers attribut is concrete.
+            # if it is *not* a hash, it is a lone LEAVE phaser, the most
+            # commonly used phaser (in the core at least).
+            my $phasers := nqp::getattr(
+              nqp::getcodeobj($coderef),Block,'$!phasers'
+            );
+
+            # slow path here
+            if nqp::ishash($phasers) {
+                my @leaves := nqp::atkey($phasers, '!LEAVE-ORDER');
+                my @posts  := nqp::atkey($phasers, 'POST');
+                my @exceptions;
+                unless nqp::isnull(@leaves) {
+                    my @keeps  := nqp::atkey($phasers, 'KEEP');
+                    my @undos  := nqp::atkey($phasers, 'UNDO');
+                    my int $n := nqp::elems(@leaves);
+                    my int $i := -1;
+
+                    # fast leave path
+                    if nqp::isnull(@leaves) && nqp::isnull(@undos) {
+                        while ++$i < $n {
+                            CATCH { nqp::push(@exceptions, $_) }
 #?if jvm
-                    nqp::decont(nqp::atpos(@leaves,0))();
+                            nqp::atpos(@leaves, $i))();
```

Comment:
There is a superfluous closing paren now (leftover from removing the ```nqp::atpos```). CI is also not happy with that.

Reply:
Thanks for the spot, fixed now!